hazelcast/src/main/java/com/hazelcast/sql/impl/client/ReadOptimizedLruCache.java

```diff
@@ -0,0 +1,148 @@
+ * access time. It allows the size to grow beyond the capacity, up to
+ * `cleanupThreshold`, at which point the inserting thread will remove a batch
+ * of the eldest items in two passes.
+ * <p>
+ * The cleanup process isn't synchronized to guarantee that the capacity is not
+ * exceeded. The cache is available during the cleanup for reads and writes. If
+ * there's a large number of writes by many threads, the one thread doing the
+ * cleanup might not be quick enough and there's no upper bound on the actual
+ * size of the cache. This is done to optimize the happy path when the keys fit
+ * into the cache.
+ */
+public class ReadOptimizedLruCache<K, V> {
+
+    // package-visible for tests
+    final ConcurrentMap<K, ValueAndTimestamp<V>> cache;
+
+    private final AtomicBoolean cleanupLock = new AtomicBoolean();
+    private final int capacity;
+    private final int cleanupThreshold;
+
+    /**
+     * @param cleanupThreshold The size at which the cache will clean up oldest
+     *     entries in batch. `cleanupThreshold - capacity` entries will be removed.
+     */
+    public ReadOptimizedLruCache(int capacity, int cleanupThreshold) {
+        if (capacity <= 0) {
+            throw new IllegalArgumentException("capacity <= 0");
+        }
+        if (cleanupThreshold <= capacity) {
+            throw new IllegalArgumentException("cleanupThreshold <= capacity");
+        }
+
+        this.capacity = capacity;
+        this.cleanupThreshold = cleanupThreshold;
+
+        cache = new ConcurrentHashMap<>(cleanupThreshold);
+    }
+
+    public V getOrDefault(K key, V defaultValue) {
+        final V existingValue = get(key);
+        return existingValue != null ? existingValue : defaultValue;
+    }
+
+    public V get(K key) {
+        ValueAndTimestamp<V> valueFromCache = cache.get(key);
+        if (valueFromCache == null) {
+            return null;
+        }
+        valueFromCache.touch();
+        return valueFromCache.value;
+    }
+
+    public void put(K key, V value) {
+        if (value == null) {
+            throw new IllegalArgumentException("Null values are disallowed");
+        }
+
+        ValueAndTimestamp<V> oldValue = cache.put(key, new ValueAndTimestamp<>(value));
+        if (oldValue == null && cache.size() > cleanupThreshold) {
+            doCleanup();
+        }
+    }
+
+    public void remove(K key) {
+        cache.remove(key);
+    }
+
+    private void doCleanup() {
+        // if no thread is cleaning up, we'll do it
+        cleanupLock.compareAndSet(false, true);
+        try {
+            int entriesToRemove = cache.size() - capacity;
+            if (entriesToRemove <= 0) {
+                // this can happen if the cache is concurrently modified
+                return;
+            }
+            PriorityQueue<Long> oldestTimestamps =
+                    new PriorityQueue<>(entriesToRemove + 1, Comparator.<Long>naturalOrder().reversed());
```

Comment:
What is the advantage of using heap here over ordering the timestamps in a list and cut it from threshold value? Since ordering will only be nlog(n) but keep pooling/adding to heap in each time will multiply the nlog(n) complexity.

Reply:
I did a little benchmark:
```java
public class MyBenchmark {

    private static final int NUM_ITEMS = 100;
    private static final int NUM_TO_REMOVE = NUM_ITEMS / 10;

    @State(Scope.Thread)
    public static class MyState {
        private List<Long> values = new Random().longs(NUM_ITEMS).boxed().collect(toList());
    }

    @Benchmark
    public void sortingWithPriorityQueue(MyState state) {
        PriorityQueue<Long> queue = new PriorityQueue<>();
        for (Long value : state.values) {
            queue.add(value);
            if (queue.size() > NUM_ITEMS / 10) {
                queue.poll();
            }
        }
    }

    @Benchmark
    public void sortingACollection(MyState state, Blackhole blackhole) {
        List<Long> sortedValues = new ArrayList<>();
        for (Long value : state.values) {
            sortedValues.add(value);
        }
        Collections.sort(sortedValues);
        blackhole.consume(sortedValues.get(NUM_TO_REMOVE));
    }
}
```
Results:
```
Benchmark                              Mode  Cnt       Score       Error  Units
MyBenchmark.sortingACollection        thrpt   25  322750.483 ± 17986.694  ops/s
MyBenchmark.sortingWithPriorityQueue  thrpt   25  464215.275 ± 22058.800  ops/s
```
I used the `PriorityQueue` somewhat intuitively. It only stores 1/10th of the data, while if we sort the collection, it has to store and sort the whole dataset. That's why the PriorityQueue is faster. It also takes less memory I think.

Reply:
I benchmarked also the case when removing 50 items out of 150 items. I expected the collection sorting to be rather comparable, but it's actually still 33% faster...

Reply:
Thank you for your detail explanation ❤️. However, I'm still confused. How do you be sure that you remove oldest values without adding all values to the queue? `if (queue.size() > NUM_ITEMS / 10) ` will be valid without seeing the rest of the values in the queue. Isn't it weird?

Reply:
The queue is sorted from newest timestamps to oldest. When queueSize is larger, we `poll` it, which removes the newest timestamp. Therefore, after the first cache iteration, the queue will contain as many _oldest_ timestamps as we need to remove. So if we poll the queue again, we'll get a timestamp, we get the `removeThreshold`.